// cardinal-direction/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const CardinalIncrement = 11.25;
function cardinalIntFromDegree(degree) {
    const int = Math.round(degree / CardinalIncrement);
    return int > 31 ? 0 : int;
}
function ensureProperType(value, block) {
    if (typeof value === "string") {
        return block[value];
    }
    else {
        return block[block[value]];
    }
}
function cardinalFromIntSubset(cardinalInt, cardinalSubset) {
    const subset = ensureProperType(cardinalSubset, CardinalSubset);
    const int = Math.round(cardinalInt / subset) * subset;
    return CardinalDirection[int > 32 - subset + subset / 2 ? 0 : int];
}
/**
 * There are 32 cardinal-type direction on a compass.
 * - This numeric enum is `0` indexed.
 */
var CardinalDirection;
(function (CardinalDirection) {
    CardinalDirection[CardinalDirection["N"] = 0] = "N";
    CardinalDirection[CardinalDirection["NbE"] = 1] = "NbE";
    CardinalDirection[CardinalDirection["NNE"] = 2] = "NNE";
    CardinalDirection[CardinalDirection["NEbN"] = 3] = "NEbN";
    CardinalDirection[CardinalDirection["NE"] = 4] = "NE";
    CardinalDirection[CardinalDirection["NEbE"] = 5] = "NEbE";
    CardinalDirection[CardinalDirection["ENE"] = 6] = "ENE";
    CardinalDirection[CardinalDirection["EbN"] = 7] = "EbN";
    CardinalDirection[CardinalDirection["E"] = 8] = "E";
    CardinalDirection[CardinalDirection["EbS"] = 9] = "EbS";
    CardinalDirection[CardinalDirection["ESE"] = 10] = "ESE";
    CardinalDirection[CardinalDirection["SEbE"] = 11] = "SEbE";
    CardinalDirection[CardinalDirection["SE"] = 12] = "SE";
    CardinalDirection[CardinalDirection["SEbS"] = 13] = "SEbS";
    CardinalDirection[CardinalDirection["SSE"] = 14] = "SSE";
    CardinalDirection[CardinalDirection["SbE"] = 15] = "SbE";
    CardinalDirection[CardinalDirection["S"] = 16] = "S";
    CardinalDirection[CardinalDirection["SbW"] = 17] = "SbW";
    CardinalDirection[CardinalDirection["SSW"] = 18] = "SSW";
    CardinalDirection[CardinalDirection["SWbS"] = 19] = "SWbS";
    CardinalDirection[CardinalDirection["SW"] = 20] = "SW";
    CardinalDirection[CardinalDirection["SWbW"] = 21] = "SWbW";
    CardinalDirection[CardinalDirection["WSW"] = 22] = "WSW";
    CardinalDirection[CardinalDirection["WbS"] = 23] = "WbS";
    CardinalDirection[CardinalDirection["W"] = 24] = "W";
    CardinalDirection[CardinalDirection["Wbs"] = 25] = "Wbs";
    CardinalDirection[CardinalDirection["WNW"] = 26] = "WNW";
    CardinalDirection[CardinalDirection["NWbW"] = 27] = "NWbW";
    CardinalDirection[CardinalDirection["NW"] = 28] = "NW";
    CardinalDirection[CardinalDirection["NWbN"] = 29] = "NWbN";
    CardinalDirection[CardinalDirection["NNW"] = 30] = "NNW";
    CardinalDirection[CardinalDirection["NbW"] = 31] = "NbW";
})(CardinalDirection = exports.CardinalDirection || (exports.CardinalDirection = {}));
/**
 * Used to restrict the type of cardinal direction that can be returned by `cardinalFromDegree`. When a subset is specified the degree will be pushed the the nearest acceptable cardinal.
 * @member Full - All directions can be returned. ___[DEFAULT]___
 * @member Basic - Specifies only basic _(N,E,S,W)_ can be returned.
 * @member Ordinal - Extends on `Basic`, allowing _(NE,SE etc..)_ to be returned.
 * @member Intercardinal - Extends on `Ordinal, allowing _(NNE,ENE etc...)_ to re returned.
 */
var CardinalSubset;
(function (CardinalSubset) {
    CardinalSubset[CardinalSubset["Full"] = 1] = "Full";
    CardinalSubset[CardinalSubset["Basic"] = 8] = "Basic";
    CardinalSubset[CardinalSubset["Ordinal"] = 4] = "Ordinal";
    CardinalSubset[CardinalSubset["Intercardinal"] = 2] = "Intercardinal";
})(CardinalSubset = exports.CardinalSubset || (exports.CardinalSubset = {}));
/**
 * Converts a compass degree [0-360] into its correlated cardinal direction.
 * @param degree - The compass degree to covert [0-360].
 * @param subset - _[optional]_ Specifies the type of cardinal that can be returned. See `CardinalSubset` enum for details.
 * @return [string] A `CardinalDirection` enum member value.
 */
function cardinalFromDegree(degree, subset) {
    const int = cardinalIntFromDegree(degree);
    return subset ? cardinalFromIntSubset(int, subset) : CardinalDirection[int];
}
exports.cardinalFromDegree = cardinalFromDegree;
/**
 * Converts a cardinal direction into its correlated compass degree value.
 * @param cardinal - One of the `CardinalDirection` enum member values.
 * @return [number] A compas degree between [0-360].
 */
function degreeFromCardinal(cardinal) {
    const cardinalInt = ensureProperType(cardinal, CardinalDirection);
    return CardinalIncrement * cardinalInt;
}
exports.degreeFromCardinal = degreeFromCardinal;
